pkg/agent/controller/networkpolicy/fqdn.go

```diff
@@ -405,80 +408,108 @@ func (f *fqdnController) deleteRuleSelectedPods(ruleID string) error {
 
 func (f *fqdnController) onDNSResponse(
 	fqdn string,
-	responseIPs map[string]net.IP,
-	lowestTTL uint32,
+	newIPWithExpiration map[string]ipWithExpiration,
 	waitCh chan error,
 ) {
-	if len(responseIPs) == 0 {
+	if len(newIPWithExpiration) == 0 {
 		klog.V(4).InfoS("FQDN was not resolved to any addresses, skip updating DNS cache", "fqdn", fqdn)
 		if waitCh != nil {
 			waitCh <- nil
 		}
 		return
 	}
-	// mustCacheResponse is only true if the FQDN is already tracked by this
-	// controller, or it matches at least one fqdnSelectorItem from the policy rules.
-	// addressUpdate is only true if there has been an update in IP addresses
-	// corresponded with the FQDN.
-	mustCacheResponse, addressUpdate := false, false
-	recordTTL := time.Now().Add(time.Duration(lowestTTL) * time.Second)
+
+	addressUpdate := false
+	currentTime := time.Now()
+	ipWithExpirationMap := make(map[string]ipWithExpiration)
+
+	// timeToRequery establishes a maximum reference time for tracking the minimum re-query time to DNS, as IPs expire.
+	timeToRequery := currentTime.Add(24 * time.Hour)
+
+	updateMapWithIP := func(ip string, ipMeta ipWithExpiration) {
+		ipWithExpirationMap[ip] = ipMeta
+		if ipMeta.expirationTime.Before(timeToRequery) {
+			timeToRequery = ipMeta.expirationTime
+		}
+	}
```

Comment:
```suggestion
	// timeToRequery establishes a maximum reference time for tracking the minimum re-query time to DNS, as IPs expire.
	var timeToRequery *time.Time

	updateMapWithIP := func(ip string, ipMeta ipWithExpiration) {
		ipWithExpirationMap[ip] = ipMeta
		if timeToRequery == nil || ipMeta.expirationTime.Before(*timeToRequery) {
			timeToRequery = &ipMeta.expirationTime
		}
	}
```
To avoid introducing the vague duration `24 * time.Hour`

Reply:
Another way is to move least expiration calculation to a function of `dnsMeta`, then there is no need to have this function, and it can save a copy in "First time seeing this domain" case

Reply:
This is awesome. I also had earlier thought on similar lines but i thought of it using as :
```go 
var timeToRequery time.Time
```
but as you know it meant ```timeToRequery``` gets set to its zero value and would not have worked because its zero value would have always been lesser, making me to set an arbitrary value to it.